LwM2M suspend and  resume api

jheiskan81 opened the pull request:
New API for suspend and resume LwM2M engine.

New event LWM2M_RD_CLIENT_EVENT_ENGINE_SUSPENDED for indicate application
that engine is suspended.

Simplify stack suspend and resume state same time for queue mode.

New CONFIG_LWM2M_RD_CLIENT_SUSPEND_SOCKET_AT_IDLE for enable skip socket close
at RX_OFF_IDLE state that socket is only suspended and close is called only
when connection is resumed.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -6178,6 +6227,21 @@ int lwm2m_engine_start(struct lwm2m_ctx *client_ctx)
+{
+	k_thread_resume(socket_loop_tid);
+	LOG_INF("LWM2M engine socket receive thread resume ");
+	
```
Unneeded empty line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -6189,7 +6253,7 @@ static int lwm2m_engine_init(const struct device *dev)
 
 	/* start sock receive thread */
-	k_thread_create(&engine_thread_data, &engine_thread_stack[0],
+	socket_loop_tid = k_thread_create(&engine_thread_data, &engine_thread_stack[0],
```
Perhaps `engine_thread_id` would be a more descriptive name?

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5833,6 +5890,9 @@ static void socket_loop(void)
 
 		for (i = 0; i < sock_nfds; i++) {
+
+			if (sock_ctx[i]->sock_fd < 0)
```
Was this change intended? It seems a bit off.

jheiskan81 replied:
yes true. I need to check. I think there should be continue inside.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1450,28 +1455,104 @@ int lwm2m_init_message(struct lwm2m_message *msg)
+
+int lwm2m_engine_connection_resume(struct lwm2m_ctx *client_ctx)
+{
+
```
Unneeded empty line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1450,28 +1455,104 @@ int lwm2m_init_message(struct lwm2m_message *msg)
+
+	if (client_ctx->sock_fd >= 0 && !client_ctx->connection_suspended) {
+		int socket_temp_id = client_ctx->sock_fd;
+
```
Redundant empty line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -55,9 +55,10 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #include <zephyr/init.h>
 #include <zephyr/sys/printk.h>
-
+#include <net/socket.h>
```
```suggestion
#include <zephyr/net/socket.h>
```

Is it really needed though here?

jheiskan81 replied:
OK I can remove and test.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1450,28 +1455,104 @@ int lwm2m_init_message(struct lwm2m_message *msg)
+	return ret;
+}
+
+int lwm2m_close_suspend(struct lwm2m_ctx *client_ctx)
```
Why `close_suspend`? Nothing seems to be closed in this case. Perhaps `lwm2m_socket_suspend()` would be enough?

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -1091,6 +1097,44 @@ static void lwm2m_rd_client_service(struct k_work *work)
+			client.engine_state = ENGINE_SUSPENDED;
+			client.suspend_thread = false;
+			k_mutex_unlock(&client.mutex);
+			while(get_sm_state() == ENGINE_SUSPENDED) {
+				k_sleep(K_MSEC(STATE_MACHINE_UPDATE_INTERVAL_MS / 2));
+			}
```
I'm a bit confused again, isn't the engine thread supposed to be suspended at this point? Why is this loop here?

jheiskan81 replied:
I change a design so don't worry.

jheiskan81 commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -197,6 +197,12 @@ config LWM2M_TLS_SESSION_CACHING
+config LWM2M_RD_CLIENT_SUSPEND_SOCKET_AT_IDLE
+	bool "Socket close is skipped at RX_ON_IDLE state"
+	help
+	  This config is not close socket when LwM2M engine enter suspend'
+	  or RX_OFF_IDLE state.
```
OK I will fix this.

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1450,28 +1455,104 @@ int lwm2m_init_message(struct lwm2m_message *msg)
+		/* Enable Queue mode buffer store */
+		client_ctx->buffer_client_messages = true;
+#endif
+		lwm2m_socket_update(client_ctx);
```
OK- let me re-think can I avoid this operation.

jheiskan81 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -6178,6 +6227,21 @@ int lwm2m_engine_start(struct lwm2m_ctx *client_ctx)
+{
+	LOG_INF("LWM2M engine socket receive thread paused");
+
+	k_thread_suspend(socket_loop_tid);
```
currently RD Client is calling this. Is calling this at own thread. API call must be call from different thread I think.

Commit:
net: lwm2m: LwM2M Pause and resume support

New API for suspend and resume LwM2M engine.
New event LWM2M_RD_CLIENT_EVENT_ENGINE_SUSPENDED for indicate
application that engine is suspended.

Simplify stack suspend and resume state same time for queue mode.

New CONFIG_LWM2M_RD_CLIENT_SUSPEND_SOCKET_AT_IDLE for enable skip socket
close at RX_OFF_IDDLE state that socket is only suspended and close is
called only when connection is resumed.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

Commit:
net: lwm2m: Shell helper for pause and resume

Helper commit for test pause and resume.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

jheiskan81 commented:
@rlubos I have been updated your raised issues and changes pause and resume functionality. @SeppoTakalo Please review.

Commit:
samples: net: lwm2m_client: Updated RD client event handler

Added missed LWM2M_RD_CLIENT_EVENT_ENGINE_SUSPENDED event handler

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

SeppoTakalo commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -197,6 +197,12 @@ config LWM2M_TLS_SESSION_CACHING
 	help
 	  Enabling this only when feature is supported in TLS library.
 
+config LWM2M_RD_CLIENT_SUSPEND_SOCKET_AT_IDLE
```
I wonder if we should drop the whole `LWM2M_RD_CLIENT` Kconfig. I don't believe the engine could work anymore without it. I also don't think anyone test or use it in that configuration. I really cannot see the point.

However, not in this PR though..

rlubos replied:
Not in this PR, but I agree that it'd make sense to drop this config.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -201,6 +198,15 @@ static void set_sm_state(uint8_t sm_state)
+
+	/* Suspend socket after Event callback */
+	if (event == LWM2M_RD_CLIENT_EVENT_QUEUE_MODE_RX_OFF) {
+#if defined(CONFIG_LWM2M_RD_CLIENT_SUSPEND_SOCKET_AT_IDLE)
+		lwm2m_socket_suspend(client.ctx);
+#else
+		lwm2m_close_socket(client.ctx);
+#endif
```
```suggestion
		if (IS_ENABLED(CONFIG_LWM2M_RD_CLIENT_SUSPEND_SOCKET_AT_IDLE)) {
				lwm2m_socket_suspend(client.ctx);
		} else { 
				lwm2m_close_socket(client.ctx);
		}

```

SeppoTakalo reviewed (dismissed):
Looking good.
Just some styling suggestions.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -161,26 +166,103 @@ char *sprint_token(const uint8_t *token, uint8_t tkl)
+
 int lwm2m_push_queued_buffers(struct lwm2m_ctx *client_ctx)
 {
+	#if defined(CONFIG_LWM2M_QUEUE_MODE_ENABLED)
```
Preprocessor code shouldn't be indented.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -192,9 +274,9 @@ int lwm2m_push_queued_buffers(struct lwm2m_ctx *client_ctx)
 		msg = SYS_SLIST_CONTAINER(msg_node, msg, node);
 		sys_slist_append(&msg->ctx->pending_sends, &msg->node);
 	}
+	#endif
```
Ditto

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -487,6 +532,18 @@ int lwm2m_socket_add(struct lwm2m_ctx *ctx)
+		sock_fds[i].fd = ctx->sock_fd;
+		return;
+	}
+
```
Redundant empty line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -433,6 +435,22 @@ static int cmd_update(const struct shell *sh, size_t argc, char **argv)
+static int cmd_pause(const struct shell *sh, size_t argc, char **argv)
+{
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
```
`sh` also seems to be unused.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -433,6 +435,22 @@ static int cmd_update(const struct shell *sh, size_t argc, char **argv)
+static int cmd_resume(const struct shell *sh, size_t argc, char **argv)
+{
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
```
`sh` also seems to be unused.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -859,6 +923,44 @@ int lwm2m_engine_start(struct lwm2m_ctx *client_ctx)
+
+	str = k_thread_state_str(engine_thread_id, buffer, sizeof(buffer));
+	if (suspend_engine_thread || !strcmp(str, "suspended")) {
+		LOG_WRN("Engine Already at Suspended");
```
Just a suggestion:
```suggestion
		LOG_WRN("Engine thread already suspended");
```

rlubos reviewed (commented):
Just a few nits.

jheiskan81 commented:
@SeppoTakalo & @rlubos Thanks for feedback. I have been fixed raised items now.
